qir/stdlib/src/stdlib.cpp

```diff
@@ -0,0 +1,137 @@
+#include <cstdint>
+#include <cstring>
+#include <utility>
+
+class Array
+{
+public:
+  Array(int64_t size_, int64_t count_) noexcept
+    : size{std::move(size_)}
+    , count{std::move(count_)}
+  {
+    data = new int8_t[size * count];
+  }
+
+  Array(Array *ref) noexcept
+    : size{ref->size}
+    , count{ref->count}
+  {
+    data = new int8_t[size * count];
+    memcpy(data, ref->data, size * count);
+  }
+
+  ~Array() noexcept
+  {
+    delete[] data;
+  }
+
+  int64_t size;
+  int64_t count;
+  int8_t *data;
+  int64_t alias_count{0};
+  int64_t ref_count{1};
+};
+
+extern "C"
+{
+  int8_t *__quantum__rt__tuple_create(int64_t n)
+  {
+    auto  ret = new int8_t[n + 3 * sizeof(int64_t)];
+    auto &s   = *reinterpret_cast<int64_t *>(ret);
+    auto &r   = *reinterpret_cast<int64_t *>(ret + sizeof(int64_t));
+    auto &a   = *reinterpret_cast<int64_t *>(ret + 2 * sizeof(int64_t));
+    s         = n;
+    r         = 1;
+    a         = 0;
+    return ret + 3 * sizeof(int64_t);
+  }
+
+  void __quantum__rt__tuple_update_reference_count(int8_t *tuple, int32_t n)
+  {
+    auto  ptr = tuple - 3 * sizeof(int64_t);
```

Comment:
What is `3`? Perhaps a named constant to clear it up?

Reply:
Yes, the entire file needs to be cleaned up.